src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs

```diff
@@ -91,8 +91,23 @@ public static unsafe SocketError CreateSocket(AddressFamily addressFamily, Socke
             return errorCode;
         }
 
+        private static unsafe int SysRead(SafeSocketHandle socket, Span<byte> buffer, out Interop.Error errno)
+        {
+            int received = 0;
```

Comment:
In SysSend/SysReceive we are asserting socket.IsSocket. Should we assert !socket.IsSocket here?

Reply:
I've added it to the methods that require a socket type file descriptor. The `Assert` helps identify the code leading to the invalid usage.

`SysRead`/`SysWrite` could be used also with socket type file descriptors, so I didn't add it there.

Reply:
> SysRead/SysWrite could be used also with socket type

They could... but in practice they aren't, right? And I'm not sure why they ever would be. Maybe I'm missing something here?

Reply:
Currently they aren't. But we may try to see if that gives a perf improvement (probably not...).

The reason I added the asserts to the socket related functions is that they clearly showed me what invalid use leads to `SocketError.NotSocket`. The `SocketException` gets thrown higher up, so it doesn't include the interesting part of the stack.

Do you want me to add the asserts to `SysSend/SysReceive` for `!socket.IsSocket`?

Reply:
Given the way this code works, I think it's useful to have those asserts, yeah. That would prevent a subsequent change from inadvertently doing something weird here.

Reply:
I've added the asserts.